astropy/units/quantity_helper/function_helpers.py

```diff
@@ -488,7 +492,7 @@ def arange_impl(*args, start=None, stop=None, step=None, dtype=None, device=None
         if v is not None
     }
 
-    new_values, out_unit = _quantities2arrays(*qty_kwargs.values(), unit_from_first=True)
+    new_values, _ = _quantities2arrays(*qty_kwargs.values())
```

Comment:
Thinking about this, I think you still need `unit_from_first`, otherwise if `stop` has no unit and `start` has one, that unit from `start` will be taken but discarded here. I think it may be best to have an explicit test that if `stop` has no unit, all other values also have no unit. That probably can then remain a separate branch, with this piece only happening for the case that `stop` has a unit.